calyx/src/passes/register_unsharing.rs

```diff
@@ -0,0 +1,287 @@
+                            self.get_cell(builder, old_name),
+                            self.get_cell(builder, new_name),
+                        ))
+                    }
+                    GroupOrInvoke::Invoke(invoke) => {
+                        invoke_map.entry(invoke.clone()).or_default().push((
+                            self.get_cell(builder, old_name),
+                            self.get_cell(builder, new_name),
+                        ))
+                    }
+                }
+            }
+        }
+
+        for (grp, rename_cells) in grp_map {
+            let group = self.get_group(builder, grp);
+            let mut group_ref = group.borrow_mut();
+            builder.rename_port_uses(&rename_cells, &mut group_ref.assignments)
+        }
+
+        self.invoke_map = invoke_map;
+    }
+}
+
+impl Visitor for RegisterUnsharing {
+    fn start(
+        &mut self,
+        comp: &mut ir::Component,
+        _c: &LibrarySignatures,
+    ) -> VisResult {
+        self.bookkeeper.replace(Bookkeeper::new(comp));
+        let mut builder = Builder::from(comp, _c, false);
+
+        let rename_list = self
+            .bookkeeper
+            .as_mut()
+            .unwrap()
+            .create_new_regs(&mut builder);
+
+        self.bookkeeper
+            .as_mut()
+            .unwrap()
+            .rename(&mut builder, &rename_list);
+
+        // for (group, z) in &self.bookkeeper.as_ref().unwrap().analysis.reach {
+        //     println!("Group {:?}", group);
+        //     println!("  {:?}", z);
+        // }
+
+        // for (x, y) in &self
+        //     .bookkeeper
+        //     .as_ref()
+        //     .unwrap()
+        //     .analysis
+        //     .calculate_overlap(&builder.component.continuous_assignments)
+        // {
+        //     println!("Overlapping defns for {}", x);
+        //     for def in y {
+        //         println!("   {:?}\n", def);
+        //     }
+        // }
+        Ok(Action::Continue)
+    }
+
+    fn invoke(
+        &mut self,
+        invoke: &mut ir::Invoke,
+        _comp: &mut ir::Component,
+        _sigs: &LibrarySignatures,
+    ) -> VisResult {
+        if let Some(name) = extract_meta_name(invoke) {
+            let vec_array =
+                &self.bookkeeper.as_ref().unwrap().invoke_map.get(&name);
+
+            // only do rewrites if there is actually rewriting to do
+            if let Some(rename_vec) = vec_array {
+                replace_invoke_ports(invoke, rename_vec);
+            }
+
+            clear_meta_name(invoke);
+        }
+
+        Ok(Action::Continue)
+    }
+}
+
+fn extract_meta_name(invoke: &ir::Invoke) -> Option<ir::Id> {
+    if let Some(counter) = invoke.attributes.get(INVOKE_PREFIX) {
+        Some(ir::Id::from(format!("{}{}", INVOKE_PREFIX, counter)))
+    } else {
+        None
+    }
+}
+
+fn clear_meta_name(invoke: &mut ir::Invoke) {
+    invoke.attributes.remove(INVOKE_PREFIX);
```

Comment:
Are these attributes added by the reaching defs analysis?

Reply:
Correct

Reply:
Resolving this for the moment since this function doesn't exist with the metadata table which disappears when the analysis is dropped.